File: POH6plus/test1.js
p=process.stdin;
p.resume();
p.setEncoding('utf8');
p.on('data',function(chunk){
    w=chunk.toString().split('\n').sort();
    n=w[1];c=s=e='';
    for(i=0;i<n;i++){
        z=w[i+2];
        r=z.split(e).reverse().join(e);
        if(z==r)c+=z;else if(w.indexOf(r)>0&&z<r)s+=z;
    }
    console.log(s+c+s.split(e).reverse().join(e));
});

File: POH6plus/test1.java
class Main {
    public static void main(String[]a)throws Exception {
        byte[]b=new byte[20000];System.in.read(b);
        String[]w=new String(b).split("\\s");
        StringBuilder x=new StringBuilder(9000),c=new StringBuilder();
        int n=Integer.parseInt(w[0]),i,j;
        java.util.Arrays.sort(w,1,n+1);
        for(i=1;i<=n;i++){
            String v=w[i],r=new StringBuilder(v).reverse().toString();
            for(j=i+1;j<=n&&!r.equals(w[j]);j++);
            if(j<=n){x.append(v);w[j]="";}else if(r.equals(v))c.append(r);
        }
        System.out.print(x);
        System.out.print(c.append(x.reverse()));
    }
}
